### ISSUE TITLE
Adding members after cluster initialization?

### ISSUE THREAD

**Issue by tchoulihan** [atomix/atomix#21]
Lets say I start up a single node, that only defines itself as a member. If I start up a 2nd node, that lists the first as its member, will the first node accept it? This wasn't clear after reading through the doc.

Or, do all member need to be initialized exactly with each other, to start up?

Also, two more quick things.
1) The snapshot isn't deployed to maven central, please run a mvn deploy to do this. 
2) I'm thinking about using this to replicate a sqlite database. Would the file event log be my best bet? IE writing the INSERTS and such to the file log. Or would I have to use the state machine resource instead?

Thanks.

edit: I was finally able to install this locally by doing a git clone, and running `mvn clean install -DskipTests`, because there were a lot of test failures.

**Comment by kuujo** [comment]
Hey,

Currently, the answer to your question about adding a member to the cluster is: no, adding a second member that knows about the first member will not result in proper behavior. Cluster membership changes for "active" (Raft voting) members is a somewhat complex topic (I'll explain why in a moment). In the current implementation nodes only know about members that are configured on startup. So if node B starts after node A, then node A needs to have been started with knowledge of node B and vice versa for replication to work. The current plan is for this behavior to remain and for membership changes to be left out of the first release (but I may be persuaded on this as I know membership changes are important) which should be coming up shortly.

The reason membership changes present a challenge is because it's not quite as simple as adding a node and replicating existing members' logs to the new member. Because Raft's consistency is dependent on quorums (Raft defines commitment of a command as being persisted in a majority of the servers in the cluster) and membership changes cause a potential change in the size of a quorum, membership change implementations must be careful to ensure that new members cannot contribute to the voting protocol before catching up with existing commits.

For instance, if you add a second node to your single node "cluster," the quorum then increases to two. If that second node immediately joins as a full voting member without first having received all the committed commands from the first node, then any commands that were previously committed are suddenly replicated on less than a majority of the voting cluster (still 1 node). Additionally, because a two-node Raft cluster requires commands to be replicated on both nodes in order to be committed, and because commands are always processed in order, any new commands submitted to the two-node cluster will be blocked or timeout until the second node's log is caught up with the first. In other words, if we define commitment as being persisted in a majority of nodes, adding nodes to the cluster can result in committed commands becoming uncommitted and new commands being blocked if not done correctly.

So, a correct implementation of membership changes requires more than just allowing configurations to be changed; there are multiple concerns. First, the Raft implementation needs to be updated to commit membership changes to the underlying log. That's fine. But secondly, additional mechanisms must be put in place to track the state of new nodes joining the cluster. One suggested method is to add membership statuses to nodes. In that case, the second node would join the cluster as a "promotable" member which is not allowed to vote. Once that member's log is caught up with the first node's commitIndex, the second node is promoted to a full voting member and the cluster is officially a two-node Raft cluster. However, this requires quite a bit more logic around handling promotions and propagating that information through the log.

It's not necessarily that membership changes are difficult to implement, but that it's critical that they be implemented correctly for the reasons stated above. For this reason, I want to ensure that the core implementation is verified first before adding membership changes.

As for the SQLite issue, this is actually a topic I've been discussing quite a bit lately - persistent state machines. I would indeed recommend you use the event log for this. The reason is that the current version's log compaction implementation (snapshots) does not work well in the context of persistent state machines. If your database is always going to be small enough to fit on a single node then you can certainly replicate it via Raft.

But one potential problem with the event log and any other Copycat log with regard to persistent state machines is that entries can and will be applied twice (after a failure). This means that persistent state machines need to be designed for idempotency. The way this can be achieved is by using the event log index as a high watermark. For instance, when writing to a SQL database, write the index of the entry in the same row as the data. This ensures that the index is persisted at the same time and in the same location as the state, allowing you to guarantee it will be applied only once to your state machine (SQLite) by performing CaS operations rather than blind inserts.

Another potential problem with the current incarnation of the event log is that it has no concept of its replication status. The event log is designed to arbitrarily drop old segments of committed entries according to a RetentionPolicy regardless of whether the entries in those segments have been replicated on 100% of the nodes in the cluster. The new version, however, solves this problem by tracking both when a log is replicated on a majority of the nodes in the cluster (committed) and when it's replicated on all the nodes in the cluster. This allows Copycat to ensure that only segments that have been seen by the entire cluster can be deleted from disk.

**Comment by tchoulihan** [comment]
Hey @kuujo , thanks a ton for your help with this. Copycat looks like the most well-maintained and feature-filled java raft implementation.

> So if node B starts after node A, then node A needs to have been started with knowledge of node B and vice versa for replication to work. The current plan is for this behavior to remain and for membership changes to be left out of the first release (but I may be persuaded on this as I know membership changes are important) which should be coming up shortly.

Ah, this is unfortunate. I'm currently using go-raft, which does have dynamic cluster features, and is essential for my purposes. I don't know all the nodes beforehand which might join and need to catch up on replication.

> The reason membership changes present a challenge is because it's not quite as simple as adding a node and replicating existing members' logs to the new member. Because Raft's consistency is dependent on quorums (Raft defines commitment of a command as being persisted in a majority of the servers in the cluster) and membership changes cause a potential change in the size of a quorum, membership change implementations must be careful to ensure that new members cannot contribute to the voting protocol before catching up with existing commits.

That makes sense. Perhaps a node status could be added, like *syncing*, where all other functions are postponed until this operation completes.

Thats good to hear that the event log would serve my purposes for syncing SQL. I'm currently using a sqlite syncing project called [rqlite](https://github.com/otoolep/rqlite), and it works great for the most part, except that its based on go-raft, which is now unmaintained and has a lot of reliability issues when nodes go down. I was hoping to use copycat to potentially build my own implementation of syncing sqlite across a lot of nodes, or a whitelist of nodes having full write power that syncs across the network.

A kind of decentrally hosted database, using the reliability of raft.

> But one potential problem with the event log and any other Copycat log with regard to persistent state machines is that entries can and will be applied twice (after a failure).

I'm sure I don't understand the problem completely, but if a failure occurred, that event has 3 possibilities: written, not written, or corrupted. 

If it wasn't written, then it knows it needs to catch up for everything just before that event. 

If it was written, then it needs to catch up for everything just after that event. 

I don't see why certain events would be written twice, unless the failure occurred exactly **while** an event was in the process of being written, in which case a corrupted event should default to being rewritten. There's not much else you can do in that case; its an unavoidable(and probably rare) problem that everyone would have to just accept. And it would be extremely rare in stable nodes where writes are fast.

> The new version, however, solves this problem by tracking both when a log is replicated on a majority of the nodes in the cluster (committed) and when it's replicated on all the nodes in the cluster. This allows Copycat to ensure that only segments that have been seen by the entire cluster can be deleted from disk.

One scary thought regarding this when you have a potentially dynamic cluster, is that all the current nodes are synced, but what if a node joins much later? If the event log was deleted, the new node wouldn't be able to catch up on the earlier actions. A helpful option would be to be able to turn off event log compaction/deleting/ or whatever its called.

**Comment by kuujo** [comment]
> On Apr 23, 2015, at 7:27 AM, Tyler Houlihan <notifications@github.com> wrote:
> 
> Hey @kuujo , thanks a ton for your help with this. Copycat looks like the most well-maintained and feature-filled java raft implementation.
> 
> So if node B starts after node A, then node A needs to have been started with knowledge of node B and vice versa for replication to work. The current plan is for this behavior to remain and for membership changes to be left out of the first release (but I may be persuaded on this as I know membership changes are important) which should be coming up shortly.
> 
> Ah, this is unfortunate. I'm currently using go-raft, which does have dynamic cluster features, and is essential for my purposes. I don't know all the nodes beforehand which might join and need to catch up on replication.
> 
> The reason membership changes present a challenge is because it's not quite as simple as adding a node and replicating existing members' logs to the new member. Because Raft's consistency is dependent on quorums (Raft defines commitment of a command as being persisted in a majority of the servers in the cluster) and membership changes cause a potential change in the size of a quorum, membership change implementations must be careful to ensure that new members cannot contribute to the voting protocol before catching up with existing commits.
> 
> That makes sense. Perhaps a node status could be added, like syncing, where all other functions are postponed until this operation completes.
> 
So, I think I can agree that this is needed. I'm currently in the midst of writing much more extensive tests than Copycat has ever had before, and I expect that it will take some time. I spent all day today just writing basic tests for all commands and their configurable persistence and consistency levels (Copycat now has per-request persistence/consistency levels, allowing for different configurations based on the command being executed). Hopefully within the next few days Copycat should be more stable than it was previously and I'll be merging all this work back in to master. 

I'll be out of town at the end of this week, but after that I think it would be worth my while to add membership changes in. It should be doable in a manner that's consistent with the existing gossip based membership. Essentially, starting a new ACTIVE member would cause it to automatically join the cluster and receive logs from other nodes, and shutting an ACTIVE member down gracefully would cause it to leave the cluster. This just take a week or so to hack together.
> Thats good to hear that the event log would serve my purposes for syncing SQL. I'm currently using a sqlite syncing project called rqlite, and it works great for the most part, except that its based on go-raft, which is now unmaintained and has a lot of reliability issues when nodes go down. I was hoping to use copycat to potentially build my own implementation of syncing sqlite across a lot of nodes, or a whitelist of nodes having full write power that syncs across the network.
> 
> A kind of decentrally hosted database, using the reliability of raft.
> 
> But one potential problem with the event log and any other Copycat log with regard to persistent state machines is that entries can and will be applied twice (after a failure).
> 
> I'm sure I don't understand the problem completely, but if a failure occurred, that event has 3 possibilities: written, not written, or corrupted.
> 
I'm not referring to entries being written to the underlying Raft log twice. That will never happen. The worst that could happen on that side is a corrupted entry or a missing index (in which case the entry will not have been committed - entries are fsynced to the log long before they're marked committed - so this is irrelevant). What I'm referring to is entries being applied to the state machine twice. 

On each node in the cluster, once an entry is committed the entry is applied to the state machine - e.g. some state method in Copycat's StateMachine or the consumer on Copycat's EventLog or the commit handler on Copycat's StateLog  - in this case a *persistent* SQL-based state machine. It's important to note that Copycat does *not* persist state regarding the last entry applied to the state machine. If the node is shut down or dies, when it restarts Copycat does not retain any information regarding which entries were applied and will therefore replay all the entries in the log again in order to rebuild the state machine's state. This works great to rebuild state for in-memory state machines (e.g. Copycat's collections), but persistent state machines must guard against this via idempotency. Without accounting for the duplicate application of entries in a persistent state machine, you will likely be writing duplicate rows in a SQL table. For instance, if you're persisting rows of names in a SQL database, and the first three names are Jack, Jill, and Jane, Copycat will call the event log consumer with Jack, Jill, and Jane when those entries are first committed and again after the next restart as well.

So, the way to create idempotency in a persistent state machine is by storing the last applied index *with the data*. One may think that Copycat could simply persist the last applied index, but there are problems with this. For instance, if Copycat were to apply an entry to the persistent state machine and then die before persisting the fact that the entry was applied, that entry will still be applied again on restart. So, because of the nature of Copycat and the separation between its internal persistence later and users' potentially persistent state machines, this type of caveat can be more safely handled on the users' side, i.e. in a user-defined consumer or state machine. By storing the index of each row in your SQL table, the last applied index is affectively flushed to disk with the data and is guaranteed to be persisted at the same time as the state update (the SQL row).
> If it wasn't written, then it knows it needs to catch up for everything just before that event.
> 
> If it was written, then it needs to catch up for everything just after that event.
> 
> I don't see why certain events would be written twice, unless the failure occurred exactly while an event was in the process of being written, in which case a corrupted event should default to being rewritten. There's not much else you can do in that case; its an unavoidable(and probably rare) problem that everyone would have to just accept. And it would be extremely rare in stable nodes where writes are fast.
> 
> The new version, however, solves this problem by tracking both when a log is replicated on a majority of the nodes in the cluster (committed) and when it's replicated on all the nodes in the cluster. This allows Copycat to ensure that only segments that have been seen by the entire cluster can be deleted from disk.
> 
> One scary thought regarding this when you have a potentially dynamic cluster, is that all the current nodes are synced, but what if a node joins much later? If the event log was deleted, the new node wouldn't be able to catch up on the earlier actions. A helpful option would be to be able to turn off event log compaction/deleting/ or whatever its called.
> 
You can turn it off. You just have to configure the retention policy with FullRetentionPolicy which never deletes segments. But the issue with this is that it's not scalable. All entries in Copycat's log will persist for eternity, so eventually it will consume your disk.

This is part of why I redesigned the log compaction algorithm - for event handling. Snapshots obviously don't fit well with the event model, and retention policies are not ideal. So, with the impending merge, all of Copycat's logs will now be key-based and perform key-based log cleaning. That means users can associate an arbitrary key with an entry in the log and Copycat's log compaction algorithm will periodically compact the log by attempting to remove *all but the last committed entry* with any given key. So, this process affords the user some control over which entries are evicted from the log. Copycat's AsyncMap uses its map keys as log keys. So, for instance, if a user calls put(key, value), get(key), put(key, value) on the key "foo" in an AsyncMap, Copycat will eventually drop the first put(key, value) from the log and won't even log the get(key) since it doesn't contribute to the map's state.

This sort of simple key-based log cleaning doesn't completely solve the problem of log compaction, though. For instance, if a user calls put(key, value), get(key), put(key, value), and remove(key), that last remove(key) entry (basically a tombstone) would persist in the log for eternity (since Copycat's log retains the last entry for each key) even though it results in an absence of state. So, ideally we could find a way to eventually drop that remove(key) entry as well since in the event of a replay it amounts to nothing more than a no-op. However, it's still essential that the remove(key) operation be applied to state machines on all nodes. If it's dropped from the log before it's applied by all nodes, then some node that still contains the state from the previous put(key, value) will retain that state and will retain the put(key, value) entry in its log.

So, the new Copycat update accounts for these types of tombstone entries. As I mentioned, Copycat allows for several persistence levels:  NONE, EPHEMERAL, DURABLE, and PERSISTENT (and also consistency levels: EVENTUAL, LEASE, and STRICT). Entries persisted with the DURABLE persistence level can be compacted from the log *only after they've been logged by all active Raft voting members*. This behavior is achieved by leaders simply tracking the highest index seen by all members and sharing that information in AppendEntries RPCs. Copycat's StateMachine assigns the DURABLE/STRICT persistence/consistency levels to all commands annotated with the @Delete annotation in order to guarantee the correct behavior.

So, that's all to say that I recommended the current event log for a persistent state machine specifically because Copycat's existing snapshot implementation is not designed for persistent state machines and indeed will not work well with them. But ideally, the final Copycat implementation could be used since it fixes many of these exact types of problems. It's going to be a few days before I get it caught up to the current master branch, but based on current development progress, I'm very confident that it will surpass that version in terms of stability and flexibility in very short order (a few days).

**Comment by kuujo** [comment]
By the way, with regard to the cluster membership change questions... I
should mention that Copycat does already support arbitrarily adding members
to the cluster. This is supported on PASSIVE members via a gossip-based
membership detection and replication system. So, when a PASSIVE member
joins the cluster, resources will be replicated to that member
asynchronously. That means the member's state will not be perfectly in-sync
with that of the Raft leader, but its state is eventually consistent.

Additionally, the new update adds a third member type - REMOTE - which
allows purely remote interaction with resources, meaning no data is
replicated to REMOTE members.

**Comment by domsj** [comment]
Interesting to see work being done here towards better supporting persistent state machines.
I had a similar use case in mind, but instead of sql I was thinking of using a graph database (tinkerpop / orientdb).

`By storing the index of each row in your SQL table, the last applied index is affectively flushed to disk with the data and is guaranteed to be persisted at the same time as the state update (the SQL row).`

Just a small remark here. This should be read as saying that each row in the database has to contain the logindex ?
Wouldn't it be better to have 1 table with just 1 row dedicated to having the latest logindex?
That seems to solve the tombstone problem you're mentioning, unless I'm misunderstanding what it's about...

**Comment by domsj** [comment]
Ok, read the part about tombstones again, apparently I missed that it's related to the key-based log cleaning...

I expected compaction and snapshotting to work together as follows:
Every once in a while a node creates a snapshot of it's state. This snapshot + all newer log entries contain all relevant state, hence all older log entries may now be deleted.
(I'm assuming the snapshot is just as persistent as the log.)
When a node needs to do a catchup from another node there are 2 possible cases:
- it can get up to date by receiving missing log entries (without introducing a gap in the log)
- it's too much behind on log entries (there would be a gap), so it first needs to receive a snapshot, followed by the latest log entries

**Comment by tchoulihan** [comment]
I know its been a while since we've discussed this, but is this any nearer a possibility? IE, adding members after copycat has been initialized?

**Comment by kuujo** [comment]
If you look on the master branch, I added some info regarding an impending beta release. 

There are a lot of changes that have gone in to this first release. As part of those changes, I've done some work to support membership changes, but they're not currently merged in to the coming release and I'm not sure they will be by June 21st(ish). The plan is to complete testing of the existing critical components - particularly the Raft implementation, log, communication, and sessions - before continuing on to complete membership changes. It's likely I'll begin working on membership changes again in parallel with the release branch, but whether it makes it in to the beta release will depend both on the stability of the existing release code and the membership changes branch over the next few weeks.

Regardless of whether they are in the June 21st(ish) release, I've ensured at least that the framework can easily support membership changes from an architectural standpoint, and I'll continue to work on them in parallel.

Truthfully, there's not anything particularly blocking membership changes aside from the need to reduce complexity during this phase of testing. Once we are confident in the existing algorithm, membership changes should be fairly trivial to implement. But I think it most likely I'll integrate them into the second release.

**Comment by arnonmoscona** [comment]
Thanks. I'll take a further look as soon as I can

**Comment by tchoulihan** [comment]
@kuujo thanks a bunch sir, good to know that this is in the works.

**Comment by kuujo** [comment]
Both snapshots and log cleaning can work independently of each other or together. I think if Copycat were using snapshots, I would probably just stick with them rather than introducing the added complexity of log cleaning. 

But there are several reasons Copycat is no longer using snapshots, the most important of which are scalability and complexity. Snapshots essentially require that large amounts of state stored on disk can be associated with specific points in the Raft log. If a snapshot is taken at index 100, then entries 1-100 need to be removed that snapshot needs to always be installed at index 100.

This is fine for small state machines, but it can pose significant issues for larger state machines for which more than a brief moment is required to write the full state machine state to disk. For these cases, Raft can support streaming the snapshot to disk. Essentially, what you do is associate specific portions of the snapshot with specific points in the log, and associate those portions of the snapshot with specific state in the state machine. For instance, in a large key-value store, one might snapshot some smaller subset of the keys at a time. However, considering that Copycat exposes an interface for arbitrary user-defined state machines rather than a single implementation specific state machine, supporting streaming snapshots for large state machines would require a significantly more complex state machine API to allow users to associate state with portions of a snapshot.

There are other alternatives to streaming snapshots for large state machines. For instance, when a node prepares to take a snapshot, if it's the leader it can transfer its leadership to another node before stepping down and blocking to take the snapshot. But this reduces the availability f the cluster during the snapshot period, and it requires not insignificant additional coordination.

Alternatively, log cleaning moves most of the added complexity into the log itself. The log is responsible for segmenting compacting itself. I've found this to significantly reduce the complexity of the Raft implementation in addition to providing some performance benefits resulting from compaction being spread over time, using efficient sequential writes to disk. There is no additional need for coordination between servers - no snapshot replication or leadership transfers - only normal Raft log replication.

It does, however, add some increased complexity to state machine implementations themselves. State machines are responsible for tracking sessions, indexes, timestamps, and other information that may be relevant to log compaction and indicating to the compaction process which commands no longer contribute to the state machine's state, but much of this can be abstracted for common uses cases such as key-value stores (using key-based de duplication).

**Comment by kuujo** [comment]
Dynamic membership changes (the original topic of this issue) have been fixed and tested in [Copycat](http://github.com/atomix/copycat) and released in beta1 for Copycat and Atomix.

**Comment by amit2103** [comment]
@kuujo  Sorry for the extremely late reply on this but can you please give an example. I could not understand how to add a member once Server has started?

**Comment by kuujo** [comment]
Hey, sure. The documentation is still catching up...

So, when you first form an Atomix cluster, typically you create something like three replicas. When creating each replica, you provide the local address and the list of all three replicas in the cluster:

```java
List<Address> cluster = Arrays.asList(
  new Address("localhost", 5000),
  new Address("localhost", 5001),
  new Address("localhost", 5002)
);
```

```java
AtomixReplica replica1 = AtomixReplica.builder(cluster.get(0), cluster)
  ...
  .build();
replica1.open();
```

```java
AtomixReplica replica2 = AtomixReplica.builder(cluster.get(1), cluster)
  ...
  .build();
replica2.open();
```

```java
AtomixReplica replica3 = AtomixReplica.builder(cluster.get(2), cluster)
  ...
  .build();
replica3.open();
```

Now, once the cluster is started, Atomix allows you to add new replicas to the cluster simply by pointing them at the existing replicas. Cluster configuration changes are done transparently to the user. Simply create a fourth replica with the existing cluster configuration, and the replica will automatically join the cluster.

```java
AtomixReplica replica4 = AtomixReplica.builder(new Address("localhost", 5004), cluster)
  ...
  .build();
replica4.open();
```

Essentially, all you have to do is start a replica at an `Address` outside of the existing cluster configuration, and that replica will automatically request to join the cluster when started. When a replica joins an existing cluster, that cluster's structure will change from, in this case, three to four nodes. So, if you were able to access the underlying `CopycatServer`, you would be able to see that the cluster has four `Member`s after starting the fourth replica via `server.cluster().members().size()`

**Comment by amit2103** [comment]
Ohhh.Cool...I am assuming this is how copycat also works...ManyThanks

**Comment by davidecaroselli** [comment]
Hi guys,

this is to a me a fundamental feature that now allows me to integrate this fantastic framework in my work!
I think many people just like me needs to dynamically add and remove nodes to the cluster: I'm thinking basically to an automatic scaled cloud cluster scenario for example.

This is a great news and, in my opinion, it worths update the documentation asap!

Many thanks!

**Comment by jhalterman** [comment]
@davidecaroselli et al, I updated the [clustering docs](http://atomix.io/atomix/docs/clustering/) to include this stuff, along with some other bits that make them generally more useful :)

**Comment by kuujo** [comment]
@amit2013 you are absolutely correct

**Comment by kuujo** [comment]
Since the docs are still being updated, I should elaborate a bit more on the idea of adding and removing nodes arbitrarily. So, at the Copycat/Raft level, adding nodes can be a fairly expensive operation. The actual configuration change will be quick, but adding a Raft node to the cluster impacts the quorum size, so in some cases some availability can be lost until the new node is caught up. For instance, if you add a second node to a single node cluster, the cluster cannot accept writes until the new node is caught up so that writes can be replicated to a majority of nodes (i.e. both nodes) in the cluster. So, any time a node is added, it has to be caught up with the leader fairly quickly. Copycat does take some steps to mitigate availability loss during this period though.

But Atomix actually builds on top of Copycat's membership change support to facilitate use cases where nodes may be continuously added to or removed from the cluster. The problem is, Raft requires that writes go to a majority of the cluster. If you want an Atomix cluster that can scale up and down arbitrarily, then you probably don't want all the nodes to participate in the Raft algorithm. Your Atomix cluster could be scaled to 50 nodes, and in that case you don't want a single write to have to be replicated to 26 of those nodes.

So, what I haven't talked about here is Atomix's quorumHint and backupCount configurations. In my example above, the cluster is started with three nodes. That sets the cluster's quorum size to 3, meaning a write to the cluster will be replicated to a majority of that cluster - 2 nodes. Now, when the fourth node is added to the cluster, it will *not* impact the quorum size. The number of Raft nodes in the cluster will remain at 3, and the cost of writes will remain 2. The fourth node will act as a backup to the other three nodes. In the event that one of the Raft nodes crashes or is partitioned from the rest of the cluster for a long period of time, Atomix will demote that Raft node and promote the new node in its place. This allows the Raft cluster to again tolerate another failure by dynamically replacing the failed node.

But there's still some cost to that configuration change. Because the extra node was not participating in the Raft algorithm, it has to be caught up with the rest of the cluster. Atomix reduces the cost of replacing failed Raft nodes with the backupCount configuration. The backup count defines the number of backup replicas per active Raft non-leader replica. If the quorumHint is 3 and the backupCount is 1, there will be 3 Raft nodes and 2 backup nodes. Backups are stateful nodes to which only committed state changes are replicated from followers asynchronously. Because they're not counted in the Raft quorum and because the leader doesn't replicate to them, backup nodes don't significantly impact the latency or throughput of writes to the cluster. And because only committed entries are replicated to backups, the replication protocol is absent consistency checks and thus is significantly faster. But because backups are typically within a heartbeat of the Raft nodes in terms of state, replacing a failed Raft node with a live backup node requires only a few heartbeats which could be on the order of milliseconds. This means all the Raft nodes in the cluster can fail, and the cluster can still operate normally as long as a majority of the nodes don't crash simultaneously. Atomix will replace failed Raft nodes with backups, and failed backups with reserves.

Replacing replicas in a cluster is done in a manner that ensures the configured quorumHint is met at all times. If the quorumHint is 3 and one of the Raft nodes crashes, a backup or reserve node will be promoted *before* the Raft node is demoted. This ensures that the cluster can always tolerate the failure of a minority of Raft nodes (defined by quorumHint).

Keep in mind, the cluster balancing algorithm in Atomix is still considered experimental. There are plenty of optimizations that still need to be made. The current algorithm is fairly naive as to which backup to promote. It's possible it could promote a backup that's far behind the Raft nodes or replace a Raft node that is only briefly partitioned. Mechanisms have been added to Copycat to aid in more intelligent decision making in the Atomix cluster, but those decisions just haven't been implemented yet.

Atomix implements cluster balancing directly on Copycat's Cluster API. For Copycat, you can manage the cluster yourself, promoting and demoting servers from the Raft cluster at will:
http://atomix.io/copycat/docs/cluster/
http://atomix.io/copycat/api/latest/io/atomix/copycat/server/cluster/Cluster.html

**Comment by amit2103** [comment]
@kuujo  Many thanks for the excellent write up.
I sit possible to build a replicated state machine on top of Atomix Cluster?
Can you please, provide me with an example ? I was actually trying to build one on top of CopyCat. However from your message it seems that  Atomix clustering is more robust solution

**Comment by jhalterman** [comment]
@amit2103 Sure. The main difference between Atomix and Copycat in terms of state machines is that Copycat clusters contain a single state machine while Atomix resources can contain numerous state machines, basically, one for each resource in the cluster. You can create an Atomix resource state machine, which is the same sort of thing as a Copycat state machine:

http://atomix.io/atomix/docs/custom-resources/

**Comment by kuujo** [comment]
@amit2103 actually, as @jhalterman said Atomix state machines are just Copycat state machines (that extend the `ResourceStateMachine` class) with user friendly APIs. Quite literally, Atomix is just a state machine that manages state machines, and resources are implemented by extending the base `AbstractResource` class which submits operations to specific managed state machines. So, you can follow the Copycat state machine documentation for that portion of the resource as well (still a work in progress):
http://atomix.io/copycat/docs/state-machine/

State machines in Atomix behave almost exactly like they do in Copycat. The difference is primarily with sessions. In Copycat, the state machine can communicate with the client through its `Session`, and session events indicate when the client session is registered/unregistered/expired/closed. In Atomix, a session represents an instance of the resource on a client. In other words, in Copycat when a client connects to the cluster its session is opened, but in Atomix when a client *creates* a resource a logical session is opened for that resource.

If you look at existing Atomix resources, all of them are implemented in precisely the same way as a user would implement a custom resource. Also, if you look at the Atomix resource tests, all of them are actually implemented on Copycat. Resource modules don't even depend on the Atomix API. This demonstrates that Copycat and Atomix state machines are interchangeable. So, you can develop your state machine for Copycat and move it to Atomix or vice-versa. Atomix just gives you the ability to manage many state machines in the same cluster.

**Comment by kuujo** [comment]
By the way, @amit2103, if you want to use existing Atomix state machines as examples (which I would recommend), the two most interesting introductions are the `DistributedValue` and `DistributedLong` state machines. They're interesting in that they use both available approaches to log compaction - incremental compaction and snapshots. Most of the other resources do incremental compaction (`release`ing or `close`ing `Commit`s when they no longer contribute to the state machine's state) since it's more efficient, but it's also more complex. Snapshotting is straightforward and recommended to start.

The `DistributedValue` state machine uses incremental compaction. When a value is overwritten, the `Commit` for the previous value is `close`d, indicating to Copycat that it can remove the operation from the log.

The `DistributedLong` state machine cannot rely on incremental compaction. The problem is, because incremental compaction requires that all commands that contribute to the state machine's state be retained in the log, a counter would require every single increment to be persisted to represent the `DistributedLong` state. If a user increments the counter a billion times, your disk will fill up. So, `DistributedLong` uses snapshots to periodically persist the 64-bit number and remove prior increments and decrements from the log.

Log compaction is handled transparently by Copycat. It's simply the users responsibility to define *how* Copycat should compact the logs, and those two resources are good examples of the two options.

I'll be working on the documentation some more, but feel free to continue commenting here, hop on chat, or email me if you need help or have any more questions.

**Comment by hyperxpro** [comment]
@kuujo Can we add members after initialization in Atomix 3?
